Integrate CORP and COEP

yutakahirano opened the pull request:
This is part of the introduction of COEP
(https://github.com/whatwg/html/pull/5454). The CORP check now takes
COEP into account. Also, responses coming from service workers
are checked.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/fetch/1030.html" title="Last updated on Jun 26, 2020, 7:14 AM UTC (ea40eaf)">Preview</a> | <a href="https://whatpr.org/fetch/1030/b228c33...ea40eaf.html" title="Last updated on Jun 26, 2020, 7:14 AM UTC (ea40eaf)">Diff</a>

yutakahirano commented:
@domenic: PTAL

I cannot make COEP related links, which I __guess__ blocked by whatwg/html#5454. Please let me know if I'm wrong.

Note:
 - The reporting logic in the CORP check in https://wicg.github.io/cross-origin-embedder-policy/ is wrong, so I fixed the logic in this PR.
 - The CORP check between the original client and the SW (described in https://wicg.github.io/cross-origin-embedder-policy/#integration-sw) is included, because IIRC @annevk wanted to avoid checking CORP in places other than the fetch spec as much as possible.

domenic commented on fetch.bs:
```diff
@@ -3143,58 +3143,144 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+<var>request</var> and a <a for=/>response</a> <var>response</var>, run these steps:</p>
+<ol>
+ <li><p>Let <var>embedder policy</var> be <var>request</var>'s <a for=request>client</a>'s
+ embedder policy.
```
This will need to cross-link

domenic commented on fetch.bs:
```diff
@@ -3143,58 +3143,144 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+  <p>If the <a>cross-origin resource policy internal check</a> with "<code>unsafe-none</code>",
+  <var>request</var> and <var>response</var> returns <b>blocked</b>, then return <b>blocked</b>.
+
+  <p class="note no-backref">This is to queue only COEP related violation reports.
```
```suggestion
  <p class="note">This is to queue only COEP-related violation reports.
```

domenic commented on fetch.bs:
```diff
@@ -3143,58 +3143,144 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+  <p class="note no-backref">This is to queue only COEP related violation reports.
+
+ <li><p>If the <a>cross-origin resource policy internal check</a> with <var>embedder policy</var>'s
+ report only value, <var>request</var> and <var>response</var> returns <b>blocked</b>, then
```
"report only value" should cross-link. I guess we should export that from HTML too.

domenic commented on fetch.bs:
```diff
@@ -3143,58 +3143,144 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+ <li><p>If the <a>cross-origin resource policy internal check</a> with <var>embedder policy</var>'s
+ report only value, <var>request</var> and <var>response</var> returns <b>blocked</b>, then
+ <a>queue a cross-origin embedder policy CORP violation report</a> with
+ <var>request</var> and <var>embedder policy</var>'s report only reporting endpoint.
```
Same for "report only reporting endpoint".

Hmm, could we make this algorithm that an "embedder policy" and a mode or something, which is "normal" or "report only"? So it doesn't have to look into the internals of the "embedder policy" struct as much? I dunno, maybe it's fine to have these algorithms so tightly coupled. But it is strange to me that they are split across different specs if so.

domenic commented on fetch.bs:
```diff
@@ -3143,58 +3143,144 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
- <li><p>If <var>request</var>'s <a for=request>mode</a> is not "<code>no-cors</code>", then return
- <b>allowed</b>.
+ <li><p>Assert: <var>request</var>'s <a for=request>mode</a> is "<code>navigate</code>" or
+ "<code>no-cors</code>".
```
This is very different than https://wicg.github.io/cross-origin-embedder-policy/#integration-fetch. I guess that is intentional?

yutakahirano replied:
Yes, please see #985. The responsibility to handle requests with modes other than "no-cors" has not been clear up until now. This change clarifies that, stating that it is caller's responsibility.

yutakahirano commented on fetch.bs:
```diff
@@ -3143,58 +3143,144 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
-  <b>allowed</b>.
-  <!-- This changes with COEP's cross-origin value. -->
+  <p class="note no-backref">Only HTML's navigate algorithm uses this check for the
+  "<code>navigate</code>" <a for=request>mode</a>, for nested navigations. [[!HTML]]
```
Thank you. I don't know the meaning of "!" here, can you explain it?

domenic replied:
See https://tabatkins.github.io/bikeshed/#biblio

yutakahirano commented on fetch.bs:
```diff
@@ -3143,58 +3143,144 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
 
- <li><p>Return <b>allowed</b>.
+    <ul class=brief>
+     <li><p><var>request</var>'s <a for=request>origin</a> is <a>schemelessly same site</a> with
```
I think I didn't change anything here. The current https://fetch.spec.whatwg.org/#cross-origin-resource-policy-check uses "schemelessly same site".

domenic replied:
I meant compared to https://wicg.github.io/cross-origin-embedder-policy/#abstract-opdef-cross-origin-resource-policy-internal-check. I think the change from Fetch -> COEP spec is intentional, but I am unsure if it is tested...

yutakahirano replied:
Ah I see. @mikewest, did you change that intentionally?

yutakahirano commented on fetch.bs:
```diff
@@ -3143,58 +3143,144 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
-   "<code>https</code>" or <var>response</var>'s <a for=response>HTTPS state</a> is
-   "<code>none</code>"
-  </ul>
+  <dl class=switch>
```
Good catch, thanks. I added a statement sanitizing _policy_ above.

annevk commented on fetch.bs:
```diff
@@ -3871,6 +3980,10 @@ optional <i>CORS-preflight flag</i>, run these steps:
+
+       <li><p><var>request</var>'s <a for=request>response tainting</a> is "<code>opaque</code>" and
+       the <a>cross-origin resource policy check</a> with <var>request</var> and <var>response</var>
+       returns <b>blocked</b>.
```
Instead of calling it twice both here and in the HTTP layer, could we call it in a step before "If actualResponse’s status is a redirect status, then:"? Then we would only have to check the response once I think, whether it came from a service worker or the network.

It seems that would miss the 401/407 scenarios, but I wonder if those should not also invoke HTTP fetch (maybe the latter shouldn't though) as they end up bypassing the service worker as currently defined. Hmm.

Either way we should document what we're doing and why, potentially just in a source comment for now.

annevk commented on fetch.bs:
```diff
@@ -3123,58 +3143,147 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
 
-<p>To perform a <dfn>cross-origin resource policy check</dfn>, given a <var>request</var> and
-<var>response</var>, run these steps:</p>
+<p>To perform a <dfn>cross-origin resource policy internal check</dfn>, given an
```
I don't think this algorithm works now we also use it for responses from a service worker as it primarily uses the request as authority, whereas the request has almost zero authority in such situations.

This is similar to the problem with the Cache API check and needs a similar solution.

yutakahirano commented:
Hi Anne, I changed the CORP check, and it now now uses response's URL list instead of request's URL list. Can you take a look again?

annevk commented on fetch.bs:
```diff
@@ -3145,58 +3148,145 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+<p>To perform a <dfn>cross-origin resource policy internal check</dfn>, given a
+<a for=/>response</a> <var>response</var>, an <a for=url>origin</a>, <var>origin</var>, an
+embedder policy value <var>embedder policy value</var> and a boolean
+<var>for navigation</var>, run these steps:
```
Since we're checking if origin, using policy X, can read a response I'd reorder the arguments a bit.

yutakahirano replied:
Done.

annevk commented on fetch.bs:
```diff
@@ -3145,58 +3148,145 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
   `<a http-header><code>Cross-Origin-Resource-Policy</code></a>` headers will have the same effect.
 
- <li><p>If <var>policy</var> is `<code>same-origin</code>`, then return <b>blocked</b>.
+ <li><p>If <var>policy</var> is neither "<code>same-origin</code>", "<code>same-site</code>" nor
```
Oxford comma before nor.

yutakahirano replied:
Done.

annevk commented on fetch.bs:
```diff
@@ -3145,58 +3148,145 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+<p>To perform a <dfn>cross-origin resource policy internal check</dfn>, given a
+<a for=/>response</a> <var>response</var>, an <a for=url>origin</a>, <var>origin</var>, an
+embedder policy value <var>embedder policy value</var> and a boolean
+<var>for navigation</var>, run these steps:
```
Please camelCase the variable names.

yutakahirano replied:
Done.

annevk commented on fetch.bs:
```diff
@@ -3145,58 +3148,145 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
   `<a http-header><code>Cross-Origin-Resource-Policy</code></a>` headers will have the same effect.
 
- <li><p>If <var>policy</var> is `<code>same-origin</code>`, then return <b>blocked</b>.
+ <li><p>If <var>policy</var> is neither "<code>same-origin</code>", "<code>same-site</code>" nor
```
You need to use backticks here as the original text did since the result of getting a header value is a byte sequence, not a string.

yutakahirano replied:
Done.

annevk commented on fetch.bs:
```diff
@@ -3145,58 +3148,145 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+<i>navigation flag</i> set, and it's always for nested navigations. Otherwise, <var>response</var>
+is either the <a for=internal>internal response</a> of an <a>opaque filtered response</a> or
+a <a for=/>response</a> which will be the <a for=internal>internal response</a> of an
+<a>opaque filtered response</a>. [[HTML]]
```
If we make HTML not invoke this when the embedder policy is "unsafe-none", wouldn't that be easier?

yutakahirano replied:
Yeah I agree that that wouldn't work with the report-only value.

annevk replied:
Can we move this note to be after the algorithm? It reads a little weird being before the list of steps.

yutakahirano replied:
Done.

yutakahirano commented:
> Not needed for this PR, but what do you think about moving the URL list copying and CORP check into the scheme fetch algorithm? We might not even need as much URL list copying for the other schemes as redirecting to them is forbidden.

I think having the CORP check (only) in the scheme fetch is not enough because of redirects.

yutakahirano commented:
Anne, are you fine with this change except for missing links defined in the HTML spec? If so I'll fix the HTML and Service Worker PRs accordingly.

annevk commented on fetch.bs:
```diff
@@ -3145,58 +3148,145 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+
+<p>To <dfn>queue a cross-origin embedder policy CORP violation report</dfn>, given a
+<a for=/>response</a> <var>response</var>, an <a for=/>environment settings object</a>
+<var>settingsObject</var>, and a <i>report only flag</i>, run these steps:
```
```suggestion
<var>settingsObject</var>, and a boolean <var>reportOnly</var>, run these steps:
```

annevk commented on fetch.bs:
```diff
@@ -3145,58 +3148,145 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
 
+<p>To perform a <dfn export>cross-origin resource policy check</dfn>, given an <a for=url>origin</a>
+origin, an <a for=/>environment settings object</a> <var>settingsObject</var>, a
+<a for=/>response</a> <var>response</var>, and an optional <i>navigation flag</i>, run these steps:
```
```suggestion
<a for=/>response</a> <var>response</var>, and an optional boolean <var>forNavigation</var>, run these steps:
```

yutakahirano commented on fetch.bs:
```diff
@@ -1749,14 +1749,17 @@ source of security bugs. Please seek security review for features that deal with
+<var>response</var>, run these steps:
 
 <ol>
+ <li><p>Assert: <var>response</var>'s <a for=response>URL list</a> is not empty.
```
Done.

yutakahirano commented on fetch.bs:
```diff
@@ -3145,58 +3148,146 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+  "<code>unsafe-none</code>", <var>response</var>, and <var>forNavigation</var> returns
+  <b>blocked</b>, then return <b>blocked</b>.
+
+  <p class="note">This is to queue only COEP-related violation reports.
```
Done.

yutakahirano commented on fetch.bs:
```diff
@@ -3145,58 +3148,146 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+<var>response</var> is either the <a for=internal>internal response</a> of an
+<a>opaque filtered response</a> or a <a for=/>response</a> which will be the
+<a for=internal>internal response</a> of an
+<a>opaque filtered response</a>. [[HTML]]
```
Done.

yutakahirano commented:
Updated: Other than addressing comments I started using the "coep" report type (without link, as it is defined in whatwg/html#5454) to address whatwg/html#5634.

```make deploy``` still shows an error when I use ```<a for="reporting">queue</a>```. Should I wait further or is there anything I should do?

annevk commented on fetch.bs:
```diff
@@ -3143,58 +3143,144 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+   </tbody>
+  </table>
+
+ <li><p><a href="https://w3c.github.io/reporting/#queue-report">Queue</a> <var>body</var> as
```
Filed https://github.com/tabatkins/bikeshed/issues/1706 as it appears it's not indexed.

annevk replied:
So it is indexed in Shepherd, but https://github.com/w3c/reporting/commit/f31f654852f7db3dcae9c01570d0cf6e533a763c did not build. I now filed https://github.com/w3c/reporting/issues/213.

annevk commented:
@yutakahirano I filed an issue for that. Could you please resolve the comments here that have been resolved? @domenic do you want to take another look?

domenic commented on fetch.bs:
```diff
@@ -3145,57 +3148,146 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+
+<ol>
+ <li><p>Let <var>endpoint</var> be <var>settingsObject</var>'s embedder policy's
+ report only reporting endpoint if the <var>reportOnly</var> is true and
```
These should cross-link to HTML, which won't work until the HTML PR is merged.

domenic commented on fetch.bs:
```diff
@@ -3145,57 +3148,146 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+  </table>
+
+ <li><p><a href="https://w3c.github.io/reporting/#queue-report">Queue</a> <var>body</var> as
+ the "<code>coep</code>"report type for <var>endpoint</var> on
```
```suggestion
 the "<code>coep</code>" report type for <var>endpoint</var> on
```

domenic commented on fetch.bs:
```diff
@@ -3145,57 +3148,146 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+ [[!REPORTING]]
+</ol>
+
+<p>To perform a <dfn export>cross-origin resource policy check</dfn>, given an <a for=url>origin</a>
```
This algorithm being separate from the internal version is a bit confusing. I might make it the first thing in this section, instead?

yutakahirano replied:
Done.

domenic commented on fetch.bs:
```diff
@@ -3145,57 +3148,146 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+<ol>
+ <li><p>Set <var>forNavigation</var> to false if it is not given.
+
+ <li><p>Let <var>embedderPolicy</var> be <var>settingsObject</var>'s embedder policy.
```
TODO cross-link when HTML is merged

domenic commented on fetch.bs:
```diff
@@ -3145,57 +3148,146 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+  "<code>unsafe-none</code>", <var>response</var>, and <var>forNavigation</var> returns
+  <b>blocked</b>, then return <b>blocked</b>.
+
+  <p class="note">This is to queue only Cross-Origin Embedder Policy violation reports.
```
I don't understand this note

yutakahirano replied:
Updated. Is it now clear?

domenic commented on fetch.bs:
```diff
@@ -3145,57 +3148,146 @@ response <a for=/>header</a> can be used to require checking a <a for=/>request<
+<p class="note no-backref">Only HTML's navigate algorithm uses this check with
+<var>forNavigation</var> set to true, and it's always for nested navigations. Otherwise,
+<var>response</var> is either the <a for=internal>internal response</a> of an
+<a>opaque filtered response</a> or a <a for=/>response</a> which will be the
```
```suggestion
<a>opaque filtered response</a>, or a <a for=/>response</a> which will be the
```

domenic commented on fetch.bs:
```diff
@@ -3586,9 +3678,8 @@ optionally with a <i>recursive flag</i>, run these steps:
-  this point, unless it came from a service worker, in which case it will only be empty if it was
-  created through <a lt="Response()" constructor><code>new Response()</code></a>.
+  <p class=note>A <a for=/>response</a>'s <a for=response>URL list</a> can be empty when it's for
+  "<code>about</code>" URLs for example.
```
This change seems a bit out of place? And the phrasing is a bit weird. If we keep it then I'd say

> A response's URL list can be empty (for example, when the response represents an `about:` URL).

yutakahirano replied:
Done.

jugglinmike replied:
The latest revision doesn't use @domenic's suggested phrasing.

yutakahirano replied:
Sorry, it seems I dropped the edit during merging. Fixed.

yutakahirano commented:
> Several TODOs around HTML cross-linking, but I'm not sure what's the latest on the sequencing of the PRs...

This PR heavily uses concepts defined in whatwg/html#5454. whatwg/html#5454 depends on two things in this PR.

 1. serialize a response URL for reporting.
 2. the modified parameters for the CORP check

It is possible to split this change to break the circular dependency, but I think it's just OK to land the HTML PR first because it seems OK to have a missing link there in terms of building.

domenic commented:
GitHub says this has conflicts with master, so is probably a good idea to rebase or merge it to make landing it easier.

jugglinmike commented on fetch.bs:
```diff
@@ -4513,13 +4614,12 @@ Range Requests</cite>. [[HTTP-RANGE]] However, this is not widely supported by b
   </ol>
 
+ <li><p>Set <var>response</var>'s <a for=response>URL list</a> to a <a for=list>clone</a> of
+ <var>httpRequest</var>'s <a for=request>URL list</a>.
```
@yutakahirano I'm having trouble understanding the implications of this change. Would you mind explaining this a little?

yutakahirano replied:
This PR changes a parameter of the CORP check. Currently it takes a request, but with this change it will take a response. This specific change is to set the correct URL list to the response so that the CORP check keeps working correctly.

jugglinmike replied:
Got it. Thank you!

annevk commented:
This looks good to me (and Domenic I think and presumably Yutaka), it has implementer support, and it has tests. Both Chrome and Firefox have implementation bugs, including for reporting. Should we file a general one on Safari for COOP and COEP?

Commit message seems reasonable (mostly deferring to HTML for information). We should probably mention that this changes "serialize a request URL for reporting".

If I understand it correctly the plan would be to merge HTML first, then wait a day, address the remaining integration issues as a result of Shepherd having new terms, and then merge this?

domenic commented:
> If I understand it correctly the plan...

Yep!

yutakahirano commented on fetch.bs:
```diff
@@ -4057,15 +4058,16 @@ optional <i>CORS-preflight flag</i>, run these steps:
-  and the <a>cross-origin resource policy check</a> with <var>request</var>'s
-  <a for=request>origin</a>, <var>request</var>'s <a for=request>client</a>, and
-  <var>actualResponse</var> returns <b>blocked</b>, then return a <a>network error</a>.
+  <p>If either <var>request</var>'s <a for=request>response tainting</a> or <var>response</var>'s
```
I changed here because the previous logic didn't work well for responses coming from the service worker.

annevk commented:
Tentative commit message:
```
Integrate CORP and COEP

This adds support for cross-origin embedding policy, which primarily is a way of enforcing the Cross-Origin-Resource-Policy header to be set on responses.

See also this HTML change which links tests and the various issues and standards efforts involved here: https://github.com/whatwg/html/pull/5454.

The earlier added "serialize a request URL for reporting" has been replaced with "serialize a response URL for reporting" as centering things around responses was found more logical.
```
Any final feedback?

domenic commented:
LGTM!
